presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcMetadataReader.java

```diff
@@ -415,69 +386,69 @@ static Slice maxStringTruncateAtFirstReplacementCharacter(Slice value)
-    }
+            // Currently, all versions of the ORC writer round trip string min and max values through java.lang.String which
+            // replaces invalid UTF-8 sequences with the unicode replacement character.  This can cause the min value to be
+            // greater than expected which can result in data sections being skipped instead of being processed. As a work around,
+            // the string stats are truncated at the first replacement character.
+            if (codePoint == REPLACEMENT_CHARACTER_CODE_POINT) {
+                break;
+            }
 
-    // returns index of first surrogateCharacter in the string -1 if no surrogate character is found
-    @VisibleForTesting
-    static int firstSurrogateCharacter(String value)
-    {
-        char[] chars = value.toCharArray();
-        for (int i = 0; i < chars.length; i++) {
-            if (chars[i] >= MIN_SURROGATE) {
-                return i;
+            // The original writer performs comparisons using java Strings to determine the minimum and maximum
+            // values. This results in weird behaviors in the presence of surrogate pairs and special characters.
+            //
+            // For example, unicode code point 0x1D403 has the following representations:
+            // UTF-16: [0xD835, 0xDC03]
+            // UTF-8: [0xF0, 0x9D, 0x90, 0x83]
+            //
+            // while code point 0xFFFD (the replacement character) has the following representations:
+            // UTF-16: [0xFFFD]
+            // UTF-8: [0xEF, 0xBF, 0xBD]
+            //
+            // when comparisons between strings containing these characters are done with Java Strings (UTF-16),
+            // 0x1D403 < 0xFFFD, but when comparisons are done using raw codepoints or UTF-8, 0x1D403 > 0xFFFD
+            //
+            // We use the following logic to ensure that we have a wider range of min-max
+            // * if a min string has a surrogate character, the min string is truncated
+            //   at the first occurrence of the surrogate character (to exclude the surrogate character)
+            // * if a max string has a surrogate character, the max string is truncated
+            //   at the first occurrence the surrogate character and 0xFF byte is appended to it.
+            if (version == ORIGINAL && codePoint > REPLACEMENT_CHARACTER_CODE_POINT) {
+                break;
             }
+
+            position += lengthOfCodePoint(codePoint);
         }
-        return -1;
-    }
 
-    @VisibleForTesting
-    static Slice concatSlices(Slice slice1, Slice slice2)
-    {
-        Slice slice = Slices.allocate(slice1.length() + slice2.length());
-        slice.setBytes(0, slice1.getBytes());
-        slice.setBytes(slice1.length(), slice2.getBytes());
-        return slice;
+        // signal to not truncate... this is necessary due to the way max truncation is written
+        return position == length ? -1 : position;
```

Comment:
I don't have any objection against this. But I don't see a necessity. If position is length, it unambiguously mean nothing need to be truncated.